lib/srv/ctx.go

```diff
@@ -587,51 +586,48 @@ func (c *ServerContext) ID() int {
 	return c.id
 }
 
-// SessionID returns the ID of the session in the context.
-func (c *ServerContext) SessionID() rsession.ID {
-	return c.ConnectionContext.GetSessionID()
-}
-
-// GetServer returns the underlying server which this context was created in.
-func (c *ServerContext) GetServer() Server {
-	return c.srv
-}
+// GetJoinParams gets join params if they are set.
+//
+// These params (env vars) are set synchronously between the "session" channel request
+// and the "shell" / "exec" channel request. Therefore, these params are only guaranteed
+// to be accurately set during and after the "shell" / "exec" channel request.
+//
+// TODO(Joerger): Rather than relying on the out-of-band env var params, we should
+// provide session params upfront as extra data in the session channel request.
```

Comment:
I don't know if we can do this without incurring an additional roundtrip of latency when talking to outdated servers (because we are going to have to use a separate channel type and fall back to the real ssh one on failure). I wish x/crypto/ssh had a way to set and read negotiated extensions at handshake time. 😔

Reply:
Is that going to be gracefully ignored by openssh sshd?